File: src/services/evolutionApi.js
/**
 * Evolution API integration — message sending & response formatting.
 */

/**
 * Send a text message via Evolution API.
 *
 * @param {string} phone   Recipient phone (e.g. "5511999999999@s.whatsapp.net")
 * @param {string} text    Message content
 * @param {string} [instanceName]  Evolution API instance name
 */
export async function sendMessage(phone, text, instanceName) {
  const baseUrl = process.env.EVOLUTION_API_URL;
  const apiKey = process.env.EVOLUTION_API_KEY;
  const instance = instanceName ?? process.env.EVOLUTION_INSTANCE_NAME;

  if (!baseUrl || !apiKey || !instance) {
    throw new Error('[EvolutionApi] Missing EVOLUTION_API_URL, EVOLUTION_API_KEY, or EVOLUTION_INSTANCE_NAME');
  }

  const url = `${baseUrl}/message/sendText/${instance}`;

  const res = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      apikey: apiKey,
    },
    body: JSON.stringify({
      number: phone,
      text,
    }),
    signal: AbortSignal.timeout(15_000),
  });

  if (!res.ok) {
    const body = await res.text();
    throw new Error(`[EvolutionApi] HTTP ${res.status}: ${body}`);
  }

  return res.json();
}

// ── Output Parser ───────────────────────────────────────────

/**
 * Format an agent response into a clean structure separating
 * the human-readable message from system metadata.
 *
 * @param {object} params
 * @param {string} params.text              Human-readable message
 * @param {string} params.opportunityLevel  "high" | "medium" | "low" | "unknown"
 * @param {string[]} params.suggestions     Upsell suggestions
 * @param {object} params.priceAnalysis     Price analysis object
 * @param {object} params.searchParams      Original search params
 * @returns {{ humanMessage: string, systemMetadata: object }}
 */
export function formatResponse({
  text,
  opportunityLevel,
  suggestions,
  priceAnalysis,
  searchParams,
}) {
  return {
    humanMessage: text,
    systemMetadata: {
      type: 'flight_search_response',
      opportunityLevel,
      suggestions,
      priceAnalysis,
      searchParams,
      timestamp: new Date().toISOString(),
    },
  };
}

/**
 * Build the human-readable text message from flight results.
 *
 * @param {object} params
 * @param {object|null} params.bestFlight
 * @param {object|null} params.cheapestFlight
 * @param {string} params.origin
 * @param {string} params.destination
 * @param {string} params.departureDate
 * @param {string} [params.returnDate]
 * @param {string} params.opportunityLevel
 * @param {string[]} params.suggestions
 * @returns {string}
 */
export function buildHumanMessage({
  bestFlight,
  cheapestFlight,
  origin,
  destination,
  departureDate,
  returnDate,
  opportunityLevel,
  suggestions,
}) {
  const lines = [];

  lines.push(`✈️ *Resultado da busca de voos*`);
  lines.push(`📍 ${origin} → ${destination}`);
  lines.push(`📅 Ida: ${departureDate}${returnDate ? ` | Volta: ${returnDate}` : ' (somente ida)'}`);
  lines.push('');

  if (bestFlight) {
    lines.push(`🏆 *Melhor Voo:*`);
    lines.push(formatFlightSummary(bestFlight));
    lines.push('');
  }

  if (cheapestFlight && cheapestFlight !== bestFlight) {
    lines.push(`💰 *Voo Mais Barato:*`);
    lines.push(formatFlightSummary(cheapestFlight));
    lines.push('');
  }

  if (!bestFlight && !cheapestFlight) {
    lines.push('⚠️ Nenhum voo encontrado para essa rota e data.');
    return lines.join('\n');
  }

  // Opportunity badge
  const badge = {
    high: '🟢 OPORTUNIDADE ALTA',
    medium: '🟡 OPORTUNIDADE MÉDIA',
    low: '🔴 OPORTUNIDADE BAIXA',
    unknown: '⚪ SEM DADOS',
  }[opportunityLevel] ?? '⚪ SEM DADOS';

  lines.push(`📊 *Análise:* ${badge}`);
  lines.push('');

  if (suggestions.length > 0) {
    lines.push('💡 *Sugestões de venda:*');
    suggestions.forEach((s) => lines.push(`  • ${s}`));
  }

  return lines.join('\n');
}

// ── Internal helpers ────────────────────────────────────────

function formatFlightSummary(flight) {
  const legs = flight.flights ?? [];
  const parts = [];

  if (flight.price != null) {
    parts.push(`💵 R$ ${flight.price}`);
  }

  if (flight.total_duration != null) {
    const h = Math.floor(flight.total_duration / 60);
    const m = flight.total_duration % 60;
    parts.push(`⏱️ ${h}h${m > 0 ? m + 'min' : ''}`);
  }

  if (legs.length > 0) {
    const airlines = [...new Set(legs.map((l) => l.airline).filter(Boolean))];
    if (airlines.length) parts.push(`🛩️ ${airlines.join(', ')}`);

    const stops = legs.length - 1;
    parts.push(stops === 0 ? 'Direto' : `${stops} escala(s)`);
  }

  return parts.join(' | ');
}

File: app/services/evolution_api.py
"""
services/evolution_api.py — Evolution API integration.

send_message()        → sends WhatsApp text via Evolution API
format_response()     → wraps agent output into { human_message, system_metadata }
build_human_message() → formats flight results as WhatsApp-ready markdown text
"""

import os
from datetime import datetime, timezone
import httpx

TIMEOUT = 15.0


# ── Sending ──────────────────────────────────────────────────

def send_message(phone: str, text: str, instance_name: str | None = None) -> dict:
    base_url = os.environ.get("EVOLUTION_API_URL")
    api_key = os.environ.get("EVOLUTION_API_KEY")
    instance = instance_name or os.environ.get("EVOLUTION_INSTANCE_NAME")

    if not base_url or not api_key or not instance:
        raise RuntimeError(
            "[EvolutionApi] Missing EVOLUTION_API_URL, EVOLUTION_API_KEY, or EVOLUTION_INSTANCE_NAME"
        )

    url = f"{base_url}/message/sendText/{instance}"

    with httpx.Client(timeout=TIMEOUT) as client:
        response = client.post(
            url,
            headers={"Content-Type": "application/json", "apikey": api_key},
            json={"number": phone, "text": text},
        )

    if response.status_code >= 400:
        raise RuntimeError(f"[EvolutionApi] HTTP {response.status_code}: {response.text[:200]}")

    return response.json()


# ── Response Formatting ──────────────────────────────────────

def format_response(
    text: str,
    opportunity_level: str = "unknown",
    suggestions: list | None = None,
    price_analysis: dict | None = None,
    search_params: dict | None = None,
) -> dict:
    return {
        "human_message": text,
        "system_metadata": {
            "type": "flight_search_response",
            "opportunity_level": opportunity_level,
            "suggestions": suggestions or [],
            "price_analysis": price_analysis or {},
            "search_params": search_params or {},
            "timestamp": datetime.now(timezone.utc).isoformat(),
        },
    }


def build_human_message(
    multi_results: list | None = None,
    destination: str = "",
    departure_date: str = "",
    return_date: str | None = None,
    best_flight: dict | None = None,
    cheapest_flight: dict | None = None,
    opportunity_level: str | None = None,
    suggestions: list | None = None,
) -> str:
    lines = [
        "🚀 *ANÁLISE TÉCNICA DE MALHA AÉREA*",
        f"🎯 Destino: *{destination.upper()}*",
        f"📅 Período: {departure_date}"
        + (f" até {return_date}" if return_date else " (Somente Ida)"),
        "────────────────────────",
    ]

    if multi_results:
        successful = [
            r for r in multi_results
            if not r.get("error") and (
                (r.get("flight_results") or {}).get("best_flight")
                or (r.get("flight_results") or {}).get("cheapest_flight")
            )
        ]

        if not successful:
            lines.append("⚠️ Nenhuma opção viável encontrada nos principais hubs para estas datas.")
            return "\n".join(lines)

        global_options = sorted(
            [
                {
                    "origin": r["origin"],
                    "price": (
                        (r["flight_results"].get("cheapest_flight") or {}).get("price")
                        or (r["flight_results"].get("best_flight") or {}).get("price")
                    ),
                    "opportunity": r.get("opportunity", {}),
                }
                for r in successful
            ],
            key=lambda x: x["price"] or float("inf"),
        )

        best_global = global_options[0]

        lines += [
            "",
            "🌟 *MELHOR OPORTUNIDADE GLOBAL*",
            f"📍 Origem: *{best_global['origin']}*",
            f"💰 Preço: {_format_currency(best_global['price'])}",
            f"📊 Status: {_emoji_badge(best_global['opportunity'].get('opportunity_level', 'unknown'))}",
            "",
            "📋 *RESUMO POR HUB (MÉDIA DE PREÇO)*",
        ]

        for res in successful:
            fr = res["flight_results"]
            p = (fr.get("cheapest_flight") or {}).get("price") or (fr.get("best_flight") or {}).get("price")
            status = _emoji_badge(res.get("opportunity", {}).get("opportunity_level", "unknown"))
            lines.append(f"{status} *{res['origin']}*: {_format_currency(p) if p else 'N/A'}")

        lines += ["", "💡 *INSIGHTS E ESTRATÉGIA DE CONVERSÃO*"]

        all_suggestions = list(dict.fromkeys(
            s for r in successful for s in (r.get("opportunity", {}).get("suggestions") or [])
        ))
        for s in all_suggestions[:4]:
            lines.append(f" • {s}")

    else:
        # Single result (legacy)
        origin = (multi_results or [{}])[0].get("origin", "Origem") if multi_results else "Origem"
        lines += ["", f"📍 Rota: *{origin}* → *{destination}*", ""]

        if best_flight:
            lines += ["🏆 *Opção Recomendada:*", _format_flight_summary(best_flight), ""]

        badge = _emoji_badge(opportunity_level or "unknown")
        lines += [f"📊 *Análise de Mercado:* {badge}", ""]

        if suggestions:
            lines.append("💡 *Direcionamento de Venda:*")
            for s in suggestions:
                lines.append(f" • {s}")

    return "\n".join(lines)


# ── Internal helpers ─────────────────────────────────────────

def _emoji_badge(level: str) -> str:
    return {
        "high": "🟢 ALTA",
        "medium": "🟡 MÉDIA",
        "low": "🔴 BAIXA",
        "unknown": "⚪ N/A",
    }.get(level, "⚪ N/A")


def _format_currency(value) -> str:
    if value is None:
        return "N/D"
    # Manual BRL formatter (avoids locale dependency in Alpine)
    try:
        val = float(value)
        formatted = f"{val:,.2f}".replace(",", "X").replace(".", ",").replace("X", ".")
        return f"R$ {formatted}"
    except (TypeError, ValueError):
        return "N/D"


def _format_flight_summary(flight: dict) -> str:
    legs = flight.get("flights") or []
    parts = []

    if flight.get("price") is not None:
        parts.append(f"💵 R$ {flight['price']}")

    duration = flight.get("total_duration")
    if duration is not None:
        h, m = divmod(int(duration), 60)
        parts.append(f"⏱️ {h}h{m}min" if m else f"⏱️ {h}h")

    if legs:
        airlines = list(dict.fromkeys(leg.get("airline") for leg in legs if leg.get("airline")))
        if airlines:
            parts.append(f"🛩️ {', '.join(airlines)}")
        stops = len(legs) - 1
        parts.append("Direto" if stops == 0 else f"{stops} escala(s)")

    return " | ".join(parts)
